docs: add howto guides for switching from the charm plugin (#1967)
canonical/charmcraft: docs/howto/charm-to-poetry.rst

.. _howto-migrate-to-poetry:

> Migrate from the ``charm`` plugin to Poetry

Reviewer: For consistency with existing docs and docs best practices, we should use the form `Charm plugin` and `Python plugin`.

===========================================

Many charms use `Poetry`_ to manage their Python projects. For these charms, Charmcraft
has a :ref:`craft_parts_poetry_plugin`. Migrating from the charm plugin provides some
benefits, such as no longer having to maintain a ``requirements.txt`` file. If the
charm to be migrated does not currently use poetry, refer to the
`Poetry documentation <https://python-poetry.org/docs/basic-usage/>`_ for instructions
on how to use poetry for a Python project.

Update ``charmcraft.yaml``
--------------------------

The first step is to update ``charmcraft.yaml`` to include the correct parts definition.
Depending on the history of a specific charm, it may not have an explicitly-included
``parts`` section determining how to build the charm. In this case, a ``parts`` section
can be created as follows:

.. code-block:: yaml

    parts:

>       my-charm:  # This can be named anything you want

Reviewer: A morsel for thought: we generally [discourage the use of comments](https://docs.ubuntu.com/styleguide/en/#code-examples-in-documentation) in bash, but I'm not sure where configs fit in, since we encourage users to copy them. As a general principle it's better to use the doc to explain things.

Author: We use comments pretty heavily in our [config file templates](https://github.com/canonical/charmcraft/blob/main/charmcraft/templates/init-simple/charmcraft.yaml.j2). In this case, the goal of the comment is to counter something that's a fairly common misconception amongst charm authors (that the part name needs to be named the plugin name or the actual charm name) without having an aside about it, since that's out of scope for the document.

        plugin: poetry
        source: .

> Ensure a recent enough version of ``pip`` and ``poetry``

Reviewer: ```suggestion
Select compatible versions of ``pip`` and ``poetry``
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The ``poetry`` plugin requires at least pip 22.3, released in October 2022. If the

Reviewer: ```suggestion
The Poetry plugin requires at least `pip 22.3
<https://pypi.org/project/pip/22.3>`_, released in October 2022. If the
```

charm's base uses an older version of pip, a newer version can be installed in the

> build environment using a dependency part. Likewise, a charm may require a newer

Reviewer: > newer

If I'm reading the code right, wouldn't it grab the latest?

> version of poetry than is available in the distribution's repositories. The following

Reviewer: ```suggestion
version of Poetry than is available in the distribution's repositories. The following
```

``parts`` section can be used in place of the section above to upgrade pip and Poetry
for charms that build on Ubuntu 22.04 or earlier:

.. code-block:: yaml
    :emphasize-lines: 2-9,11

    parts:
      poetry-deps:
        plugin: nil
        build-packages:
          - curl
        override-build: |
          /usr/bin/python3 -m pip install pip==24.2
          curl -sSL https://install.python-poetry.org | python3 -
          ln -sf $HOME/.local/bin/poetry /usr/local/bin/poetry
      my-charm:  # This can be named anything you want
        after: [poetry-deps]
        plugin: poetry
        source: .

Add optional dependency groups
------------------------------

If the charm has optional `dependency groups`_ that should be included when creating
the virtual environment, the ``poetry-with`` key can be used to include those groups
when creating the virtual environment.

.. note::
    This is useful and encouraged, though not mandatory, for keeping track of

>     library dependencies (see below). For an example, see

Reviewer: ```suggestion
    library dependencies, as covered in the next section. For an example, see
```

    `postgresql-operator`_.

Include charm library dependencies
----------------------------------

Unlike the ``charm`` plugin, the ``poetry`` plugin does not install the dependencies
for included charmlibs. If any of the charm libraries used have PYDEPS, these will
need to be added to the charm's dependencies, potentially as their own

> `dependency group <dependency groups>`_.

Reviewer: Unspecified link.

Author: Good catch - fixed!

To find these dependencies, check each library file for its ``PYDEPS``. A command
that can find these is::

    find lib -name "*.py" -exec awk '/PYDEPS = \[/,/\]/' {} +

If run from the base directory of a charm, this will show all the PYDEPS declarations
from all loaded charm libs.

Include extra files
-------------------

> A ``poetry`` plugin only includes the contents of the ``src`` and ``lib`` directories

Reviewer: ```suggestion
A Poetry plugin only includes the contents of the ``src`` and ``lib`` directories
```
